### ISSUE TITLE
How were the COT and GCOT constructed in the training set of CLEVR?

### ISSUE THREAD

**Issue by tgyy1995** [shikras/shikra#16]
Thank you for your wonderful work!

In section 6.1, it is mentioned that "We train our Shikra-7B (without pre-training) on CLEVR in three settings: 1) Only use Question and Answer (Q→A); 2) Use Question, CoT, and answer (Q→CA); 3) Use GCoT with Center Point annotation and answer (Q→CPointA)." 
How were the COT and GCOT constructed in the training set?

**Comment by aixiaodewugege** [comment]
+1

**Comment by gulucaptain** [comment]
i'm also confused about this content.

**Comment by kq-chen** [comment]
For the CoT method, we convert the `functional program` annotation directly using handwritten rules. You can find more information about these rules [here](https://github.com/shikras/shikra/blob/892268111996c22752c5d9711d9e60d898553c35/mllm/dataset/single_image_dataset/clevr.py#L87).

In the case of GCoT, we extend the original CLEVR `functional program` annotation with intermediate results. Additionally, we append the intermediate object location for certain steps.

Here's an example from the training set:

Image:
![image-20230717151328810 (1)](https://github.com/shikras/shikra/assets/136600500/73b9d589-6cf0-4b07-838d-e276431fdfdb)

Question: How many other things are there of the same shape as the tiny cyan matte object?

Functional program:
```json
"program": [
    {"inputs": [], "function": "scene", "value_inputs": [], "ans": [0,1,2,3,4,5]},
    {"inputs": [0], "function": "filter_size", "value_inputs": ["small"], "ans": [2,4]},
    {"inputs": [1], "function": "filter_color", "value_inputs": ["cyan"], "ans": [2]},
    {"inputs": [2], "function": "filter_material", "value_inputs": ["rubber"], "ans": [2]},
    {"inputs": [3], "function": "unique", "value_inputs": [], "ans": [2]},
    {"inputs": [4], "function": "same_shape", "value_inputs": [], "ans": [0,4]},
    {"inputs": [5], "function": "count", "value_inputs": [], "ans": 2}
  ]
```

the CoT output will be:

```
scene -> filter_size:small[0] -> filter_color:cyan[1] -> filter_material:rubber[2] -> unique[3] -> same_shape[4] -> count[5] The answer is 2.
```

the GCoT output will be:

```
scene -> filter_size:small[0] -> filter_color:cyan[1] -> filter_material:rubber[2] -> unique[3][0.665,0.504] -> same_shape[4][0.560,0.350;0.400,0.577] -> count[5] The answer is 2.
```

In the GCoT output, the normalized point coordinates [0.665,0.504] and [0.560,0.350;0.400,0.577] are highlighted in the image.

I hope this helps clarify the process. If you have any further questions, please let me know.
